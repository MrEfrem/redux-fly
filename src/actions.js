export const UPDATE_STATE = '@@COMPONENT_SET_STATE'
export const RESET_STATE = '@@COMPONENT_RESET_STATE'
