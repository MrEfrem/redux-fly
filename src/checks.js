import isPlainObject from 'lodash/isPlainObject'

/**
 * Check mountPath
 * @param  {string} path
 * @return {void}
 */
export const checkMountPath = (path) => {
  if (typeof path !== 'string' || !path.length) {
    throw new Error('Mount path must be string')
  }
}

/**
 * Check initialState
 * @param  {Object} initialState
 * @return {void}
 */
export const checkInitialState = (initialState) => {
  if (!isPlainObject(initialState)) {
    throw new Error('InitialState must be plain object')
  }
}

/**
 * Check listenActions
 * @param  {Object} listenActions
 * @return {void}
 */
export const checkListenActions = (listenActions) => {
  if (listenActions && !isPlainObject(listenActions)) {
    throw new Error('ListenActions must be plain object')
  }
}

/**
 * Check options
 * @param  {Object} options
 * @return {void}
 */
export const checkOptions = (options) => {
  if (!isPlainObject(options)) {
    throw new Error('Options must be plain object ')
  }
}

/**
 * Check properties of an object options
 * @param  {Object}  options
 * @return {void}
 */
export const checkDetailOptions = (options) => {
  if (typeof options.connectToStore !== 'boolean') {
    throw new Error('connectToStore property of an object options must be boolean')
  }
  if (typeof options.unregisterInUnmount !== 'boolean') {
    throw new Error('Property unregisterInUnmount in third parameter shall be boolean')
  }
}
