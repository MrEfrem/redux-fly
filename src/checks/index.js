import isPlainObject from 'lodash/isPlainObject'

/**
 * Check mountPath
 * @param  {string} path
 * @return {void}
 */
export const checkMountPath = (path) => {
  if (typeof path !== 'string' || !path.length) {
    throw new Error('Mount path must be string')
  }
}

/**
 * Check initialState
 * @param  {Object} initialState
 * @return {void}
 */
export const checkInitialState = (initialState) => {
  if (!isPlainObject(initialState)) {
    throw new Error('InitialState must be plain object')
  }
}

/**
 * Check options
 * @param  {Object} options
 * @return {void}
 */
export const checkOptions = (options) => {
  if (!isPlainObject(options)) {
    throw new Error('Options must be plain object ')
  }
}
