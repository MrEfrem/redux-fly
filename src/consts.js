// @flow
export const RESET_STATE = '@@COMPONENT_RESET_STATE'
export const NEW_STATE = 'newState'
export const MOUNT_PATH = '@@MOUNT_PATH'
