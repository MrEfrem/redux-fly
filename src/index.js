export { default as createReducer } from './createReducer'
export { default as createStore } from './createStore'
export { default as getState } from './getState'
