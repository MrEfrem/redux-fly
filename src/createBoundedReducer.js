import { UPDATE_STATE, RESET_STATE } from './actions'

// Create reducer bounded on instance
export default (instance, initialState, listenActions) =>
  (state = initialState, action) => {
    if ((action.instance && typeof action.instance === 'string' && action.instance === instance &&
      [UPDATE_STATE, RESET_STATE].indexOf(action.type) !== -1) || // Owner actions
      ([UPDATE_STATE, RESET_STATE].indexOf(action.type) === -1 &&
      action.type in listenActions) // Other actions
    ) {
      const reducerMap = {
        [UPDATE_STATE]: (state1, action1) => ({
          ...state1,
          ...action1.newState,
        }),
        [RESET_STATE]: () => ({
          ...initialState,
        }),
        ...listenActions,
      }
      const reducer = reducerMap[action.type]
      return reducer ? reducer(state, action) : state
    }
    return state
  }
