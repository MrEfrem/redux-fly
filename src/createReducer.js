import React from 'react'
import { connect } from 'react-redux'
import storeShape from './utils/storeShape'
import { checkMountPath, checkInitialState, checkListenActions, checkOptions, checkDetailOptions }
  from './checks'
import createBoundedReducer from './createBoundedReducer'
import setReduxState from './setReduxState'
import { RESET_STATE } from './actions'
import getState from './getState'

/**
 * Create/mount reducer
 * @param  {string} mountPath
 * @param  {Object} initialState
 * @param  {Object} listenActions
 * @param  {Object} options
 * @return {function} HOC
 */
export default (
  mountPath,
  initialState,
  listenActions,
  options = {}
) => {
  checkMountPath(mountPath)
  checkInitialState(initialState)
  checkListenActions(listenActions)
  checkOptions(options)
  const _options = {
    connectToStore: true,
    unregisterInUnmount: false,
    ...options
  }
  checkDetailOptions(_options)

  const { connectToStore, unregisterInUnmount } = _options

  const reducer = {
    [mountPath]: createBoundedReducer(mountPath, initialState, listenActions || {}),
  }

  return (WrappedComponent) => {
    let ChildComponent = WrappedComponent
    if (connectToStore) {
      ChildComponent = connect(getState(mountPath))(WrappedComponent)
    }
    return class RegisterReducer extends React.Component {
      static contextTypes = {
        store: storeShape,
      }

      constructor(props, context) {
        super(props, context)
        const { store } = context
        // Binding setReduxState with redux store
        this.setReduxState = setReduxState(mountPath, store.dispatch, store.getState)
      }

      componentWillMount() {
        const { store } = this.context
        // Registration of created reducer
        store.registerReducers(reducer)
      }

      componentWillUnmount() {
        if (unregisterInUnmount) {
          const { store } = this.context
          store.dispatch({
            type: RESET_STATE,
            instance: mountPath,
          })
          store.unregisterReducers(reducer)
        }
      }

      render() {
        return (
          <ChildComponent
            {...this.props}
            setReduxState={this.setReduxState}
          />
        )
      }
    }
  }
}
