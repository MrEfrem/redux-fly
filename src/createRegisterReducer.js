// @flow
import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import storeShape from './utils/storeShape'
import createBoundedReducer from './createBoundedReducer'
import setReduxState from './setReduxState'
import { RESET_STATE, MOUNT_PATH, UUID } from './consts'
import getState from './getState'
import genUUIDv4 from './genUUIDv4'

/**
 * Create and mount reducer
 * @param  {string} mountPath
 * @param  {Object} preloadedState
 * @param  {Object} listenActions
 * @param  {Object} options
 * @param {Object} wrapped React component
 * @return {Object} React component
 */
export default function createRegisterReducer(mountPath: string, preloadedState: Object, listenActions?: Object, options: Object,
  WrappedComponent: any
) {
  const { connectToStore, persist, actionPrefix } = options
  let ChildComponent = WrappedComponent
  if (connectToStore) {
    ChildComponent = connect((state) =>
      ({ reduxState: getState(mountPath)(state) })
    )(WrappedComponent)
  }
  const uuid = genUUIDv4()
  return class RegisterReducer extends React.Component {
    static contextTypes = {
      store: process.env.NODE_ENV === 'test' ? PropTypes.object : storeShape
    }

    reducer: ?Object
    setReduxState: any
    resetReduxState: any
    batchActions: any

    componentWillMount() {
      const { store } = this.context
      if (typeof store !== 'object') {
        throw new Error('Redux store must be created')
      }
      if (typeof store.registerReducers !== 'function') {
        throw new Error('Redux store must be enhanced with redux-fly')
      }
      this.reducer = {
        [mountPath]: createBoundedReducer(uuid, mountPath, preloadedState, listenActions || {}),
      }
      // Registration of created reducer
      store.registerReducers(this.reducer)
      // Binding setReduxState with redux store
      this.setReduxState = setReduxState(uuid, mountPath, store.dispatch, store.getState, actionPrefix)
      // Action creator RESET redux state
      this.resetReduxState = () => store.dispatch({
        type: RESET_STATE,
        [MOUNT_PATH]: mountPath,
        [UUID]: uuid
      })
    }

    componentWillUnmount() {
      // If component isn't persist then RESET redux state
      if (!persist) {
        this.resetReduxState()
      }
      this.reducer = null
      this.setReduxState = null
      this.resetReduxState = null
      this.batchActions = null
    }

    render() {
      return (
        <ChildComponent
          {...this.props}
          setReduxState={this.setReduxState}
          resetReduxState={this.resetReduxState}
          reduxMountedPath={mountPath}
          persist={process.env.NODE_ENV === 'test' ? persist : null}
        />
      )
    }
  }
}
