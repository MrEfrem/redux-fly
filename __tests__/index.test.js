import * as exports from '../src'

test('Test library exports', () => {
  expect(exports).toMatchSnapshot()
})
