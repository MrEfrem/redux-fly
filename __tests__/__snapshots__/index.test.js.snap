exports[`test Test library exports 1`] = `
Object {
  "createReducer": [Function],
  "createStore": [Function],
  "getState": [Function],
}
`;
